runtime/compiler/control/rossa.cpp

```diff
@@ -315,23 +316,53 @@ j9jit_testarossa_err(
          }
       else
          {
-         // Async compilation is disabled or recompilations triggered from jitted code
+         // Async compilation is disabled or recompilations triggered from
+         // jitted code
          //
-         // Check if the method has already been scheduled for compilation and return
-         // abruptly if so. This will reduce contention on the optimizationPlanMonitor
-         // when a profiled compilation is followed by a very slow recompilation
-         // (until the recompilation is over, the java threads will try to trigger
-         // recompilations again and again. See defect 193193)
+         // Check if the method has already been scheduled for compilation and
+         // return abruptly if so. This will reduce contention on the
+         // optimizationPlanMonitor when a profiled compilation is followed by
+         // a very slow recompilation (until the recompilation is over, the java
+         // threads will try to trigger recompilations again and again. See
+         // defect 193193)
          //
          if (linkageInfo->isBeingCompiled())
             {
             TR_J9VMBase *fe = TR_J9VMBase::get(jitConfig, vmThread);
             if (fe->isAsyncCompilation())
-               return 0; // early return because the method is queued for compilation
+               {
+               // Also check if recompilation has occurred. If it has, then this
+               // request is coming from the old body whose start PC should have
+               // already been patched to jump to the helper that forwards
+               // execution to the new body. The only way to end up here is if
+               // the helper could not determine the start PC of the new body;
+               // this can happen if the new body was invalidated, and the sync
+               // recompilation triggered by the invalidation failed.
+               //
+               // In that case, trigger a sync compilation. If it succeeds, all
+               // new invocations will get forwarded to the new body; if it
+               // fails, the startPC will get patched to revert to the
+               // interpreter. Furthermore, this also prevents contention on the
+               // optimizationPlanMonitor.
+               //
+               // See OpenJ9:20750 for details.
+               //
+               if (linkageInfo->hasBeenRecompiled())
+                  {
```

Comment:
What prevents the sync compilation to fail, revert to interpreter and the entire cycle to start all over?

Reply:
For context in the following answer, the sequence of events is:
1. foo gets compiled (original body)
2. foo gets recompiled (recomp body), foo above gets patched to jump to helper to forward execution
3. foo gets invalidated, so the recompiled body in step 2 gets patched to do a sync compile and, failing that, to revert to the interpreter.

The hang only happens because the startPC of the original body was patched not to revert to the interpreter but to jump to the helper to forward execution to the new body; even after invalidation, the original body does is untouched.

If the sync compilation fails, it patches the startPC of the original body to revert to the interpreter, but now, regardless of what happens to the `extra`, the original body's startPC will never result in `j9jit_testarossa_err` getting called. So the hang is no longer possible.

However, you're right, on failure, we're not going to patch the caller, so a direct call will continue calling the old body, which will revert to interpreter causing the extra to get set to `0x1`, triggering a comp that is very likely to fail.

Reply:
I spoke to Nathan offline, and he said because the ramclass persistence work isn't enabled by default yet, I can just revert the relevant change to `initializeMethodRunAddressNoHook`, which will also fix the hang. So will close this and open that PR heh.